### PR TITLE
WT-5680 On evicting an empty page, mark it DELETED not DISK

### PR DESCRIPTION
**Pull request by michaelcahill**
When a remove becomes stable, stop writing keys even if some history is saved.  Then when a page becomes empty, it won't have any content and will have a NULL address.  These pages should be evicted into the DELETED state.  There was left over code for LIMBO pages (where a page could be empty in a checkpoint but have relevant history): this commit removes that code.

### CODE REVIEW

File: src/reconcile/rec_visibility.c [hunk 1]
```diff
@@ -152,16 +152,16 @@ __rec_append_orig_value(
...
+  uint64_t flags, WT_PAGE *page)
 {
     /*
      * Save updates for any reconciliation that doesn't involve history store (in-memory database
      * and fixed length column store), except when the maximum timestamp and txnid are globally
      * visible.
      */
     if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX)
-        return (!__wt_txn_visible_all(session, max_txn, max_ts));
+        return (!__wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts));
```
Comment (quchenhao) [inline src/reconcile/rec_visibility.c]: Why we only need to check the visibility for on page update here? I don't think this line of change is correct. Here we want to make sure every update on the update chain is visible to all so that we can safely discard the update chain.

May be change it to
```
if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX) {
        if (has_newer_updates)
               return true;
        return (!__wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts));
}
```
before this section?

Comment (michaelcahill) [reply src/reconcile/rec_visibility.c]: Thanks @quchenhao, I think you're right.  I found this function hard to follow, so I've tried to simplify it in ba67e79: can you please take a look and let me know if this addresses your concerns?

Happy to discuss if it isn't clear or I've misunderstood.

File: src/reconcile/rec_visibility.c [hunk 2]
```diff
@@ -152,29 +152,26 @@ __rec_append_orig_value(
  *     Return if we need to save the update chain
  */
 static bool
-__rec_need_save_upd(WT_SESSION_IMPL *session, WT_UPDATE *selected_upd, uint64_t max_txn,
-  wt_timestamp_t max_ts, bool has_newer_updates, uint64_t flags, WT_PAGE *page)
+__rec_need_save_upd(
+  WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_UPDATE_SELECT *upd_select, bool has_newer_updates)
 {
+    if (F_ISSET(r, WT_REC_EVICT) && has_newer_updates)
```
Comment (quchenhao) [inline src/reconcile/rec_visibility.c]: There is a case in `evict_file` where we don't set the `WT_REC_HS` flag for eviction. In the previous versions, this would fire an assertion complaining we shouldn't have saved updates when we discard the page. So I have to put the section:
```
if (!LF_ISSET(WT_REC_HS))	
        return false;
```
before
```
if (LF_ISSET(WT_REC_EVICT) && has_newer_updates)	
        return true;
```
to handle that.
I think something have changed and it is not firing anymore, which is good.

File: src/reconcile/rec_visibility.c [hunk 3]
```diff
@@ -152,29 +152,26 @@ __rec_append_orig_value(
...
      */
-    if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX)
-        return (!__wt_txn_visible_all(session, max_txn, max_ts));
-
-    if (!LF_ISSET(WT_REC_HS))
-        return false;
-
-    if (LF_ISSET(WT_REC_EVICT) && has_newer_updates)
-        return true;
+    if (!F_ISSET(r, WT_REC_HS) && !F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type != WT_PAGE_COL_FIX)
```
Comment (quchenhao) [inline src/reconcile/rec_visibility.c]: I think this is not correct. Even there is no newer updates for in memory and fix length column store, we cannot discard the update chain if there is update that is not globally visible on the update chain. The current implementation for them are very memory inefficient as they don't support history store so we have to keep all the history in memory until they are all globally visible.

**Comment by quchenhao** [comment]
There is a test failure I think is not relevant to the change but worth checking.

File: src/reconcile/rec_visibility.c (hunk 3 above)
Comment (michaelcahill) [reply src/reconcile/rec_visibility.c]: Right, for in-memory or column store we don't return here and instead fall through to the `__wt_txn_visible_all` check at line 173.  If there are newer updates, we've already returned true at line 159.  If there are no newer updates, the on-page version is the newest and we're checking its visibility.

Are you worried about some out-of-order case (i.e., from a lower isolation level), where even though the on-page version will be seen by all readers, there is some update further along the list that is not yet visible to everyone?  If so, why is that a problem?

Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: Oh right. I misunderstood the code. This is correct. Sorry for that.

Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: I confused this line with if (!F_ISSET(r, WT_REC_HS) && !F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type != WT_PAGE_COL_FIX)

 if (!F_ISSET(r, WT_REC_HS) && F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type = WT_PAGE_COL_FIX)

**Comment by michaelcahill** [comment]
Thanks @quchenhao, the test failure is WT-5746.  As far as I can see, it's unrelated to these changes.
